src/PIL/ImageGrab.py

```diff
@@ -140,7 +140,31 @@ def grabclipboard():
                 return BmpImagePlugin.DibImageFile(data)
         return None
     else:
-        if shutil.which("wl-paste"):
+        if shutil.which("loginctl"):
+            try:
+                loginctl = subprocess.check_output("loginctl").decode().split("\n")
+            except subprocess.CalledProcessError:
+                loginctl = None
+        else:
+            loginctl = None
+
+        if loginctl is not None:
+            username = os.getlogin()
+            sessionid = [
+                line.split()[0] for line in loginctl if username in line.split()
+            ][0]
+            sessiontype = (
+                subprocess.check_output(
+                    ["loginctl", "show-session", sessionid, "-p", "Type"]
+                )
+                .decode()
+                .strip("\n")
+                .split("=")[1]
+            )
```

Comment:
```suggestion
        if os.getenv("WAYLAND_DISPLAY"):
            sessiontype = "wayland"
        elif os.getenv("DISPLAY"):
            sessiontype = "x11"
```

That would change to something like this

Reply:
Sounds reasonable to me.